api/saml/parser.py

```diff
@@ -0,0 +1,357 @@
+
+        slo_service = None
+        slo_nodes = OneLogin_Saml2_Utils.query(
+            provider_node,
+            "./md:SingleLogoutService[@Binding='%s']" % required_slo_binding.value
+        )
+        if len(slo_nodes) > 0:
+            slo_url = slo_nodes[0].get('Location', None)
+            slo_service = Service(slo_url, required_slo_binding)
+
+        signing_certificate_nodes = OneLogin_Saml2_Utils.query(
+            provider_node,
+            './md:KeyDescriptor[not(contains(@use, "encryption"))]/ds:KeyInfo/ds:X509Data/ds:X509Certificate')
+        signing_certificates = self._parse_certificates(signing_certificate_nodes)
+
+        encryption_certificate_nodes = OneLogin_Saml2_Utils.query(
+            provider_node,
+            './md:KeyDescriptor[not(contains(@use, "signing"))]/ds:KeyInfo/ds:X509Data/ds:X509Certificate')
+        encryption_certificates = self._parse_certificates(encryption_certificate_nodes)
+
+        idp = IdentityProviderMetadata(
+            entity_id,
+            ui_info,
+            name_id_format,
+            sso_service,
+            slo_service,
+            want_authn_requests_signed,
+            signing_certificates,
+            encryption_certificates)
+
+        return idp
+
+    def _parse_sp_metadata(
+            self,
+            provider_node,
+            entity_id,
+            ui_info,
+            required_acs_binding=Binding.HTTP_POST):
+        """Parses SPSSODescriptor node and translates it into a ServiceProvider object
+
+        :param provider_node: SPSSODescriptor node containing SP metadata
+        :param provider_node: defusedxml.lxml.RestrictedElement
+
+        :param entity_id: String containing IdP's entityID
+        :type entity_id: string
+
+        :param ui_info: UIInfo object containing IdP's description
+        :type ui_info: UIInfo
+
+        :param required_acs_binding: Required binding for Assertion Consumer Service (HTTP-Redirect by default)
+        :type required_acs_binding: Binding
+
+        :return: ServiceProvider containing SP metadata
+        :rtype: ServiceProvider
+
+        :raise: MetadataParsingError
+        """
+        authn_requests_signed = provider_node.get('AuthnRequestsSigned', False)
+        want_assertions_signed = provider_node.get('WantAssertionsSigned', False)
+
+        name_id_format = NameIDFormat.UNSPECIFIED.value
+        name_id_format_nodes = OneLogin_Saml2_Utils.query(provider_node, './ md:NameIDFormat')
+        if len(name_id_format_nodes) > 0:
+            name_id_format = OneLogin_Saml2_Utils.element_text(name_id_format_nodes[0])
+
+        acs_service = None
+        acs_service_nodes = OneLogin_Saml2_Utils.query(
+            provider_node,
+            "./md:AssertionConsumerService[@Binding='%s']" % required_acs_binding.value
+        )
+        if len(acs_service_nodes) > 0:
+            acs_url = acs_service_nodes[0].get('Location', None)
```

Comment:
I have a couple more "what if there's more than one?" type questions about this stuff. Does it matter?

Reply:
AssertionConsumerService has "isDefault" and "index" attributes, I added additional logic for fetching the default ACS with given attributes and taking the one with the smallest index as a fallback logic